src/transform/src/semijoin_idempotence.rs

```diff
@@ -47,45 +69,108 @@ impl crate::Transform for SemijoinIdempotence {
         relation: &mut MirRelationExpr,
         _: TransformArgs,
     ) -> Result<(), crate::TransformError> {
-        {
-            // Iteratively descend the expression, at each node attempting to simplify a join.
-            let mut let_replacements = BTreeMap::<Id, Vec<Replacement>>::new();
-            let mut gets_behind_gets = BTreeMap::<Id, Vec<(Id, Vec<MirScalarExpr>)>>::new();
-            let mut worklist = vec![&mut *relation];
-            while let Some(expr) = worklist.pop() {
-                match expr {
-                    MirRelationExpr::Let { id, value, .. } => {
-                        let_replacements.insert(
-                            Id::Local(*id),
-                            list_replacements(&*value, &let_replacements, &gets_behind_gets),
-                        );
-                        gets_behind_gets
-                            .insert(Id::Local(*id), as_filtered_get(value, &gets_behind_gets));
+        let mut let_replacements = BTreeMap::<LocalId, Vec<Replacement>>::new();
+        let mut gets_behind_gets = BTreeMap::<LocalId, Vec<(Id, Vec<MirScalarExpr>)>>::new();
+        self.action(relation, &mut let_replacements, &mut gets_behind_gets)?;
+
+        mz_repr::explain::trace_plan(&*relation);
+        Ok(())
+    }
+}
+
+impl SemijoinIdempotence {
+    fn action(
+        &self,
+        expr: &mut MirRelationExpr,
+        let_replacements: &mut BTreeMap<LocalId, Vec<Replacement>>,
+        gets_behind_gets: &mut BTreeMap<LocalId, Vec<(Id, Vec<MirScalarExpr>)>>,
+    ) -> Result<(), crate::TransformError> {
+        // At each node, either gather info about Let bindings or attempt to simplify a join.
+        self.checked_recur(|_| {
+            match expr {
+                MirRelationExpr::Let { id, value, body } => {
+                    let_replacements.insert(
+                        *id,
+                        list_replacements(&*value, let_replacements, gets_behind_gets),
+                    );
+                    gets_behind_gets.insert(*id, as_filtered_get(value, gets_behind_gets));
+                    self.action(value, let_replacements, gets_behind_gets)?;
```

Comment:
Can we perform expirations here similar to the `LetRec` clause or somehow reason that these `do_expirations` calls will always be a NoOp (in that case maybe add a comment)?

Reply:
I added some comments. No need to do expirations here, as there is only one CTE (and it can't be recursive). Note that `expirations` is local to one `LetRec`, because a `LetRec` can't reference something that is defined in an inner `LetRec`, so a definition in an inner `LetRec` can't expire something from an outer `LetRec`.